Refactored bundler completely

tallytalwar opened the pull request:
The import process for bundler is less hacky and looks very similar to how JS and ES does scene importing.
- bundler now does a dfs traversal of the scene imports
    - overrides apt yaml nodes with absolute path from the associated
    imported scene
- once a unified node representing all the imports is created
    - it goes and collects valid assets to be imported
- the list of valid assets is appended with urls of all the imported yaml files
- with this list of files handy, creates the zip archive
- Adds ability to export a unified scene where all yaml importing and node merging is already done.
  - adds an option to export a json file instead of yaml when doing unified bundling.

tallytalwar committed:
Refactored bundler completely

- bundler now does a dfs traversal of the scene imports
    - overrides apt yaml nodes with absolute path from the associated
    imported scene
- once a unified node representing all the imports is created
    - it goes and collects valid assets to be imported
- with this list of files handy, creates the zip archive

tallytalwar commented:
I have tested on a few scene files, seems to be working. Since its a complete refactor of how bundler was working will definitely like to have a thorough review for this :D.

tallytalwar committed:
Cleanup

- consistent camelCase
- update code comments

tallytalwar committed:
bump version to 0.7

matteblair commented on tangram_bundler/__init__.py:
```diff
@@ -1,97 +1,231 @@
+        if 'import' in yamlFile:
+            importNode = yamlFile['import']
+            if (type(importNode) is str):
+                importPath = os.path.abspath(directory + '/' + importNode)
```
Is `os.path.join` a better way to do this? We should probably be using `os.sep` instead of `'/'` at least.

matteblair replied:
Oh or does `urljoin` work for this?

tallytalwar replied:
Ohh this is an oversight, will clean this. Old code was using `/` everywhere!

matteblair commented on tangram_bundler/__init__.py:
```diff
@@ -1,97 +1,231 @@
         zipf = zipfile.ZipFile(filename+'.zip', 'w', zipfile.ZIP_DEFLATED)
         for file in files:
-            zipf.write(file)
+            if (ignoreFileForBundling(file) and os.path.exists(file)):
```
So if `ignoreFileForBundling` returns _true_ then we _do_ bundle the file? Seems like the function returns the opposite of what its name suggests.

tallytalwar replied:
Yeah, should have been caught in my previous PR :P

matteblair commented on tangram_bundler/__init__.py:
```diff
@@ -1,97 +1,231 @@
+            importNode = yamlFile['import']
+            if (type(importNode) is str):
+                importPath = os.path.abspath(directory + '/' + importNode)
+                subImports[importPath] = yaml.safe_load(open(importPath)) 
```
It's usually a good idea to wrap file access in a `with` block to make sure it gets closed, e.g.
```python
with open(importPath, 'r') as importFile
    result = yaml.safe_load(importFile)
```
We could refactor these calls into a function that does this.

tallytalwar replied:
👍 will update

matteblair commented on tangram_bundler/__init__.py:
```diff
@@ -1,97 +1,231 @@
+def bundler(fullFilename):
+    print fullFilename, os.getcwd()
+    filename, fileExtension = os.path.splitext(fullFilename)
+    if fileExtension == '.yaml':
```
Should probably at least accept `.yml` too. Strictly speaking we don't need to care about the file extension at all.

tallytalwar replied:
Makes sense.

matteblair commented on tangram_bundler/__init__.py:
```diff
@@ -1,97 +1,231 @@
+        print files
 
         print "Bundling ",filename,"width",len(files),"dependencies, into ",filename+".zip"
         zipf = zipfile.ZipFile(filename+'.zip', 'w', zipfile.ZIP_DEFLATED)
```
Should be easy to add logic here to skip deflate on files that are already known to be compressed (like PNGs and JPGs). I'm liking this :)

matteblair reviewed (approved):
Seems to be working well!

I have some suggestions for further clean-up and improvement.

tallytalwar committed:
Addressed review comments

tallytalwar commented:
Thanks @matteblair. Incorporated all the suggestions.

nvkelso commented:
The output should be **scenefile.zip** not `scenefile.yaml.zip`.

nvkelso commented on tangram_bundler/__init__.py:
```diff
@@ -1,100 +1,233 @@
+    print files
+
+    print "Bundling ",filename,"width",len(files),"dependencies, into ",filename+".zip"
+    zipf = zipfile.ZipFile(filename+'.zip', 'w', zipfile.ZIP_DEFLATED)
```
`filename` here needs to strip off the extension and then add zip

tallytalwar replied:
Ahh, I did not realize usage of extension was also to get the final zip file name. Will fix.

tallytalwar replied:
Should be fixed by 49aff91

nvkelso commented:
I'm also getting an error when using this new version on Bubble Wrap:

> $ tangram-bundle bubble-wrap-style.yaml 

```
bubble-wrap-style.yaml /Users/nathaniel.kelso/git-repos/bubble-wrap
Traceback (most recent call last):
  File "/usr/local/bin/tangram-bundle", line 11, in <module>
    load_entry_point('tangram-bundler==0.7', 'console_scripts', 'tangram-bundle')()
  File "build/bdist.macosx-10.13-x86_64/egg/tangram_bundler/__init__.py", line 234, in main
  File "build/bdist.macosx-10.13-x86_64/egg/tangram_bundler/__init__.py", line 209, in bundler
  File "build/bdist.macosx-10.13-x86_64/egg/tangram_bundler/__init__.py", line 92, in collectAllImports
  File "build/bdist.macosx-10.13-x86_64/egg/tangram_bundler/__init__.py", line 75, in loadYaml
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/__init__.py", line 93, in safe_load
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/__init__.py", line 71, in load
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/constructor.py", line 37, in get_single_data
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 36, in get_single_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 55, in compose_document
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 84, in compose_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 133, in compose_mapping_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 84, in compose_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 133, in compose_mapping_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 84, in compose_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 133, in compose_mapping_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 84, in compose_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 133, in compose_mapping_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 84, in compose_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 133, in compose_mapping_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 84, in compose_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 133, in compose_mapping_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 84, in compose_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 133, in compose_mapping_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 82, in compose_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 111, in compose_sequence_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 84, in compose_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 133, in compose_mapping_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 82, in compose_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/composer.py", line 110, in compose_sequence_node
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/parser.py", line 98, in check_event
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/parser.py", line 474, in parse_flow_sequence_first_entry
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/parser.py", line 477, in parse_flow_sequence_entry
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/scanner.py", line 116, in check_token
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/scanner.py", line 252, in fetch_more_tokens
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/scanner.py", line 676, in fetch_plain
  File "build/bdist.macosx-10.13-x86_64/egg/yaml/scanner.py", line 1305, in scan_plain
yaml.scanner.ScannerError: while scanning a plain scalar
  in "/Users/nathaniel.kelso/git-repos/bubble-wrap/themes/bubble-wrap-icons-shields-international.yaml", line 1438, column 41
found unexpected ':'
  in "/Users/nathaniel.kelso/git-repos/bubble-wrap/themes/bubble-wrap-icons-shields-international.yaml", line 1438, column 43
Please check http://pyyaml.org/wiki/YAMLColonInFlowContext for details.
```

nvkelso commented:
It's for lines like:
> filter: { network: [CA:BC, CA:BC:primary, CA:BC:trunk] }

It's happier with (adding the quote marks):
> filter: { network: ['CA:BC', 'CA:BC:primary', 'CA:BC:trunk'] }

Since the quote marks are optional in Tangram JS I assume they're optional in ES and we shouldn't require them in the bundler either?

tallytalwar commented:
That stacktrace for the error is in the python yaml package itself. So bubble-wrap will need to be updated with the quotation fixes.

tallytalwar committed:
zipFilename fixed

matteblair commented:
Those strings with embedded colons should definitely be quoted. It's strictly permissible in the YAML spec to have unquoted scalars containing `:`, so Tangram's parser handles it, but it's uncommon and likely to break a lot of tools.

nvkelso commented:
Okay, I'll cleanup in Bubble Wrap and retest.

nvkelso commented:
Using https://github.com/tangrams/bubble-wrap/tree/nvkelso/quote-scalers it now bundles...

But it looks like you took out the special logic around theme handling (where it should would include ALL themes regardless of their use or not). Please add a boolean to add that THEME support back as it's critical for the published bundles versus the aggressively optimized ones. Really to be backwards compatible it should be a CLI argument that turns on this new aggressive mode ;)

(Note if the theme is optionally included, it's resources need to be optionally included, too.)

tallytalwar commented:
Lets discuss this IRL. Maybe I am not understanding the requirement here, it seems odd to include all themes (irrespective of their usage or not), because bundler won't bring in the assets associated with these un-used themes. But lets talk IRL :D

matteblair commented:
Oh so you want the bundler to include all yaml files in the `themes` directory whether or not they're used? I guess that's handy for building a package of all the style options for a release like the ones hosted on Nextzen. But is this the best way to distribute the styles? If your app just pulls the scene bundle from the Nextzen URL then you're downloading about 2x the data that you actually need.

nvkelso commented:
In most cases the Mapzen basemap theme variants are actually quite small YAML only files (maybe with 1 or two textures) so I think it's more like a 1% file size difference not 100%.

matteblair commented:
The theme yaml files themselves aren't large, but they can reference image files or fonts and then we have to add those to the bundle too. Probably best to add it as an option and see what sizes come out the other end.

tallytalwar commented:
> Probably best to add it as an option and see what sizes come out the other end.

The logic in this refactor won't include any assets from other un-used themes though.

matteblair commented:
Seems like we want to iterate through files in a folder called "themes" adjacent to the given scene file, then add any yaml files we find to our list of files to import, then proceed as usual. Does that work?

tallytalwar commented:
Nopes, depending on import order certain assets might get ignored.

In this rearch, I am constructing a yaml node based on the imports and only including assets which are evaluated after all the imports are done. That was the whole point of this!

tallytalwar commented:
If we have to make it work, a root node will be created for every file in the theme directory and then it will be able to include assets from these themes.

matteblair commented:
Yeah I see what you mean - kinda cumbersome. Maybe just have a conditional branch that does the old, dumb logic?

nvkelso commented:
The shell script (over [there](https://github.com/nextzen/carto-s3-bucket/blob/master/collect.sh)) separately bundles the resources needed for the style default and then each theme gets it's own bundle. So as long as we prove that the resulting ZIPs are equivalent in this branch and master we should be fine to merge this PR.

Sorry for a confusing comment in https://github.com/tangrams/bundler/pull/10#issuecomment-427481114!

tallytalwar committed:
Fix paths during importing

allImports is created with fully resolved and normalized paths, but
during recursiveImporting it was not being resolved and normalized, so
key miss was happening

Example tron was failing

nvkelso commented:
With https://github.com/tangrams/bundler/pull/10/commits/77bb139ce0006c997a9cd4a86909ad926b6298f8 we're ready to test against all the Mapzen basemap styles (including Bubble Wrap mods in https://github.com/tangrams/bubble-wrap/tree/nvkelso/quote-scalers).

tallytalwar commented:
Let me know how your tests go with the Carto repo. Everything looked good to me. Once done we should be able to merge this.

bcamper commented on tangram_bundler/__init__.py:
```diff
@@ -1,100 +1,235 @@
+                    fileList.append(os.path.relpath(prop['url'], basePath))
+                # newer Tangram font support
+                if 'weight' in prop and type(prop['weight']) is str:
+                    fileList.append(os.path.relpath(prop['weight'], basePath))
```
I'm not clear on what's supposed to be happening here. When `fonts` is a single object, a single `url` will specify a font file to always be used for that family. When `fonts` is an **array of objects**, each will specify a different `url` for a different combination of `weight`, `style`, etc. parameters.

But the value of `weight` would never be added to a file list...?

Note the comments also don't make sense, it's not "initial" vs. "newer" support, it's just whether the user is providing one vs. multiple faces per family.

bcamper replied:
You can see the equivalent JS font path parsing code here: https://github.com/tangrams/tangram/blob/master/src/scene_loader.js#L128-L138

tallytalwar replied:
This did look suspicious to me, but I left it according to how it was done before including the comment: https://github.com/tangrams/bundler/blob/4bd7821e4c29aef41bc5a91afe5ac0ba1e0b3cc6/tangram_bundler/__init__.py#L40-L43

tallytalwar replied:
I should confirm but I also do not remember ES parser supporting urls for `weight`

bcamper replied:
On further investigation, it looks like this code was handling *multiple* faces per font definition (value is an array of face objects), but not *single* face definitions, such as in https://github.com/tangrams/tangram/blob/master/demos/scene.yaml#L38-L39:

```
fonts:
    Montserrat:
        url: fonts/montserrat.woff
```

Pushed a fix in 53e8072a01fb1111a65b611b9b6a57610e9a9b91

bcamper commented on tangram_bundler/__init__.py:
```diff
@@ -1,100 +1,235 @@
-        if ('textures' in yaml_file) and (key in yaml_file['textures']):
-            referenceTexturePath = basePath + '/' + yaml_file['textures'][ yaml_file['styles'][styleName]['shaders']['uniforms'][uniformName] ]['url']
-        explicitUniformTexturePath = basePath + '/' + yaml_file['styles'][styleName]['shaders']['uniforms'][uniformName]
+        if ('textures' in rootNode) and (key in rootNode['textures']):
```
We also support arrays of textures in uniforms, e.g.

```
shaders:
  uniforms:
    textures: [textureA, textureB]
```

tallytalwar replied:
Correct, will fix this. Thanks

tallytalwar replied:
Addressed: 7a4cb9e

Found another bug with uniform texture sequence during import path resolution. 7a4cb9e also fixes this.

bcamper commented on tangram_bundler/__init__.py:
```diff
@@ -1,100 +1,235 @@
+                        propNode = materialNode[prop]
+                        if (type(propNode) is dict and 'texture' in propNode):
+                            fileList.append(os.path.relpath(propNode['texture'], basePath))
+
```
Since Tangram JS [v0.14.0](https://github.com/tangrams/tangram/releases/tag/v0.14.0), textures can also be specified inside `draw` groups for `points` and `lines`. Those should be accounted for here too.

tallytalwar replied:
Ahh, we do not support this in ES, we should create an issue for this on ES. Will rectify, thanks.

tallytalwar replied:
Interesting, since we do not have this logic in ES, I never thought through about this, but it seems like will have to go through all layers' all draw rules to check for `texture` draw rule. I know its too late in the discussion since JS has been supporting it for a while, but will it not make sense to restrict usage of these draw rule textures to "named" textures already defined in the scene's texture block?

matteblair replied:
It looks like we did add support for this in Tangram ES https://github.com/tangrams/tangram-es/issues/1645

tallytalwar replied:
Hmm I see, but in that case we are not doing it correctly in the importer, as its these texture urls are not considered for resolution with respect to the apt imported scene. Refer: https://github.com/tangrams/tangram-es/blob/e150471ade602e0f1e4f6d80e876937eb6ac05e5/core/src/scene/importer.cpp#L291

bcamper replied:
Is there a follow-up issue for this for ES?

tallytalwar replied:
There is not, will add one.

bcamper commented on tangram_bundler/__init__.py:
```diff
@@ -1,100 +1,235 @@
+        fileList.append(referenceTexturePath)
 
+# Fetch all asset dependencies in the constructed rootNode
+def fetchDependencies(fileList, rootNode, basePath):
```
Does this logic account for absolute `http(s)://` paths in addition to relative ones?

bcamper replied:
There should also be handling for cases where a texture URL field actually has a `global.` value -- in Tangram JS, the decision was to defer parsing of these until the entire scene is merged (since their final value can't be known until all imports are complete and merged), and then they are resolved against the base scene's path. I think these would have to just be omitted entirely from a server-side bundling step?

tallytalwar replied:
Will include the logic for `texture` URLs having global values and resolve it against the base scene URL.
With respect to `https://` path resolution, based on the logic here it should, but will test it and confirm.

bcamper commented:
Regarding inclusion of "theme" files... I'm unclear why we would want to include them in zip bundle, since (if I recall) there is currently no way to load them from "outside" the zip if they aren't already referenced within it. Some syntax to "switch" between or optionally import certain files would be useful (we just haven't found what it would be yet), but as of now, I'm not clear how they can be used. Am I missing something?

tallytalwar commented:
@bcamper for themes, yeah that is what the consensus was. As I pointed above bundler should not be bundling all the themes, it should just follow the import order and incorporate the themes being imported.

tallytalwar commented:
A generic comment in light of the reviews here, bundler did need a lot of love, the old approach seemed quite hacky and loose, resulting in missing assets or inclusion of unused assets in some scenarios.

bcamper commented:
Oh, yes, URLs for weight is 100% wrong.


On Thu, Oct 18, 2018 at 12:14 AM Varun <notifications@github.com> wrote:

> *@tallytalwar* commented on this pull request.
> ------------------------------
>
> In tangram_bundler/__init__.py
> <https://github.com/tangrams/bundler/pull/10#discussion_r226164052>:
>
> > -
> -            # newer Tangram font support
> -            for weight in yaml_file['fonts'][font]:
> -                if 'url' in weight:
> -                    file_list.append(os.path.normpath(folder + '/' + weight['url']))
> +    if 'fonts' in rootNode:
> +        fontsNode = rootNode['fonts']
> +        for font in fontsNode:
> +            fontNode = fontsNode[font]
> +            for prop in fontNode:
> +                # initial Tangram font support
> +                if 'url' in prop:
> +                    fileList.append(os.path.relpath(prop['url'], basePath))
> +                # newer Tangram font support
> +                if 'weight' in prop and type(prop['weight']) is str:
> +                    fileList.append(os.path.relpath(prop['weight'], basePath))
>
> I should confirm but I also do not remember ES parser supporting urls for
> weight
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/tangrams/bundler/pull/10#discussion_r226164052>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABBXSIZsP02UwIMdVRv8yzEnwkPxT3qks5umAA9gaJpZM4XFVmD>
> .
>

bcamper committed:
fix handling of single font faces

tallytalwar committed:
Fix uniform texture sequence resolution and appending

tallytalwar commented:
Maybe there should be an option here to export the "imported" root yaml object and export that to a single yaml file.

Will save on redoing all this importing in ES/JS again!

bcamper commented:
Yes I have more thoughts on pre-merging scenes... I am still not convinced bundling into a zip is the best way to go either, but... probably a separate discussion than this PR :) Improvements here are still good and useful!

tallytalwar committed:
Add textures from styles default drawrules

tallytalwar committed:
Append textures defined in layers->drawRules

tallytalwar committed:
Resolve texture urls from draw rules

tallytalwar commented:
Addressed all the comments here:
1. texture urls from draw rules from within the default style draw and from layers draw blocks: 
e5b989c, b88d520, c0428d3
2. resolve textures with global references with the base scene file path: d90b92f

tallytalwar commented:
@bcamper PTAL, all your comments should be addressed here.

I am experimenting more with single "all imported" yaml/json. Can discuss results on slack and we can go from there.

bcamper commented on tangram_bundler/__init__.py:
```diff
@@ -1,35 +1,62 @@
+
+def appendLayerDrawRuleTextures(fileList, layerNode, basePath, globalsNode):
+    for key in layerNode:
+        if key == 'data' or key == 'filter' or key == 'visible' or key == 'enabled':
```
Might want to move these to a list in more of a "config" section where they will be less buried and easier to spot if/when new reserved words are added.

bcamper commented on tangram_bundler/__init__.py:
```diff
@@ -184,6 +224,22 @@ def resolveSceneFontsUrl(fontsNode, basePath):
 
+def resolveLayersDrawTexture(layerNode, basePath):
+    for key in layerNode:
+        if key == 'data' or key == 'filter' or key == 'visible' or key == 'enabled':
```
See note on moving to a single config list of reserved words?

bcamper commented:
@tallytalwar works for me, though now that I think about it, I wonder if textures referenced in `global` should just be left *un-resolved* against any base path? I think the idea here was that they then resolve against the base path for wherever the scene file gets loaded from; if they get resolved as part of the bundler, these paths get "baked in". (I think using texture names in globals is somewhat problematic / unpredictable anyway for this reason, but, this is the logic we settled on when it came up during complex scene authorship with themes etc.)

bcamper commented:
@tallytalwar follow-up to comment above. Refill uses this `sdk_terrain_texture` global, which gets re-defined in theme files: https://github.com/tangrams/refill-style/blob/a4484a2c07b023738128926bb598607e6de0d1af/themes/color-zinc.yaml#L83

These theme files then get loaded alongside the zip bundle, and the texture is resolved at load time... I think that's what happens? :)

e.g. I have a Refill demo with terrain that loads via:

```
                import: [
                    'https://www.nextzen.org/carto/refill-style/10/refill-style.zip',
                    'https://www.nextzen.org/carto/refill-style/10/themes/label-10.zip',
                    'https://www.nextzen.org/carto/refill-style/10/themes/color-blue.zip',
                    'https://www.nextzen.org/carto/refill-style/10/themes/no-texture.zip',
                    'https://www.nextzen.org/carto/refill-style/10/themes/terrain-shading-dark.zip'
                ]
```

tallytalwar commented:
> I think using texture names in globals is somewhat problematic

Fully agree. I still feel we should not allow texture paths in globals. There is currently no use case of it right now and previous bundler code would have not imported/resolved these paths aptly so hence no regression also.
The correct way to use globals with textures is to use the named scene texture, example:

base scene yaml:
```
# default for base scene, can be overiden by a theme
global:
    style_name_texture_name: named_texture

textures:
    named_texture:
        url: blah.png

styles:
    style_name:
        base: points
        texture: global.style_name_texture_name
```

theme yaml:
```
global:
    style_name_texture_name: theme_named_texture

textures:
    theme_named_texture:
        url: blah2.png
```

So I feel its just better to avoid explicit texture urls in globals

tallytalwar commented:
@bcamper I am not sure I follow the usecase you described here: https://github.com/tangrams/bundler/pull/10#issuecomment-433648352

The way things are wired in refill in that example, these will aptly override the textures defined in the `textures` block during scene merging and has nothing to do with globals.

tallytalwar commented:
And following up to https://github.com/tangrams/bundler/pull/10#issuecomment-434378295

I would propose taking out d90b92f commit which handles resolving urls (texture or any asset) defined in globals.

bcamper commented:
> @bcamper I am not sure I follow the usecase you described here: [#10 (comment)](https://github.com/tangrams/bundler/pull/10#issuecomment-433648352)
> 
> The way things are wired in refill in that example, these will aptly override the textures defined in the `textures` block during scene merging and has nothing to do with globals.

Oh sorry I thought that `textures` block was under the `global` block, but it is not (I still feel like I've seen that use somewhere... maybe not tough).

bcamper commented:
> And following up to [#10 (comment)](https://github.com/tangrams/bundler/pull/10#issuecomment-434378295)
> 
> I would propose taking out [d90b92f](https://github.com/tangrams/bundler/commit/d90b92f55bfb355979531b6df14352d58502ebbc) commit which handles resolving urls (texture or any asset) defined in globals.

Agree, let's remove d90b92f and move ahead with this PR otherwise.

tallytalwar committed:
Use const list for layers keys we need to ignore

tallytalwar commented:
Great, we are all on the same page 👍 

- Took d90b92f out
- Used a const list for keywords (keys in layers we need to ignore for texture asset lookup).

With this I roll the ball to @nvkelso to try on the carto files. I am able to bundle all mapzen house styles and see no issues when trying locally.

nvkelso commented:
👀

tallytalwar committed:
More cleanup and steps towards unified bundler

- update paths within rootYaml dictionary to be relative to the baseUrl
when collecting all the assets post all merging
- do not bundle texture assets which are not urls (which do not exists)
        - example won't modify texture references which use globals or
        scene texture names

tallytalwar commented:
Ok, I am almost done working on the unified bundler where ONLY ONE yaml gets bundled, which includes all merged and imported yaml files. Its all working, just adding some options for bundler usage.

tallytalwar commented:
I promise to stop at this and will do json exporting in subsequent PR :D.

tallytalwar committed:
Optional parameter to do unified bundling

tallytalwar commented:
@bcamper can you give the unified bundler a shot with JS. Works great on ES. Refer: 
e295a40

tallytalwar commented:
And with this I stop adding more features to this PR. Let me know how your testing goes @nvkelso

tallytalwar committed:
Add ability to export the unified bundled scenefile as json

tallytalwar commented:
fea3400: Was trivial enough to add an option to use json exported unified bundle. Note can only be specified when doing unified bundling.

tallytalwar commented:
@bcamper Will it be trivial to try out using a JSON scene file in tangram-js?

I have started working on a version for tangram-es to use rapidjson to import json file.

bcamper commented:
Yep, you can already load a JSON file in Tangram JS (it's also a subset of
YAML anyway :)  The end goal would be to make the YAML parser optional /
loaded at runtime, so the core library size is smaller (related, same for
zip support). But even just avoiding the scene merging overhead in the
meantime is beneficial.


On Wed, Oct 31, 2018 at 7:07 PM Varun <notifications@github.com> wrote:

> @bcamper <https://github.com/bcamper> Will it be trivial to try out using
> a JSON scene file in tangram-js?
>
> I have started working on a version for tangram-es to use rapidjson to
> import json file.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tangrams/bundler/pull/10#issuecomment-434877843>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABBXUSQ30B12eDOO80LcDh6Rd73Hc3Tks5uqi0ogaJpZM4XFVmD>
> .
>

tallytalwar commented:
Ok please approve this soon now, before I keep finding more things here to add or improve :D.

bcamper commented:
I tried testing the new options, but when I run bundler with `-u -j`, I get a zip file with all the assets, but *outside* the zip I get a file called `unifiedBundledScene.json`...?

bcamper commented:
Also seeing at least one issue with bundled Bubble Wrap - will follow up with details this eve.

tallytalwar commented:
Ahh I missed a line of code for Json usecase. Easy fix, will push in a bit.

Will try in Bubble wrap also.

On Thu, Nov 1, 2018, 9:06 AM Brett Camper <notifications@github.com wrote:

> I tried testing the new options, but when I run bundler with -u -j, I get
> a zip file with all the assets, but *outside* the zip I get a file called
> unifiedBundledScene.json...?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tangrams/bundler/pull/10#issuecomment-435086732>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAWAwRClxHYp3AF-So_zN3iVRIdm9E6Bks5uqxwggaJpZM4XFVmD>
> .
>

bcamper commented:
Seeing same filename issue with YAML mode.

On Thu, Nov 1, 2018 at 12:40 PM Varun <notifications@github.com> wrote:

> Ahh I missed a line of code for Jon usecase. Easy fix, will push in a bit.
>
> Will try in Bubble wrap also.
>
> On Thu, Nov 1, 2018, 9:06 AM Brett Camper <notifications@github.com wrote:
>
> > I tried testing the new options, but when I run bundler with -u -j, I get
> > a zip file with all the assets, but *outside* the zip I get a file called
> > unifiedBundledScene.json...?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/tangrams/bundler/pull/10#issuecomment-435086732>,
> or mute
> > the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AAWAwRClxHYp3AF-So_zN3iVRIdm9E6Bks5uqxwggaJpZM4XFVmD
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tangrams/bundler/pull/10#issuecomment-435098573>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABBXRjPWGx2PAUhlTZ_4_1IRfQVUMuHks5uqyP8gaJpZM4XFVmD>
> .
>

tallytalwar committed:
Make sure json file is bundled in the zip bundle

tallytalwar commented:
Yeah that file is created, but should also be present in the zip bundle. I can delete the file after bundling it, for brevity sakes.

tallytalwar commented:
After d7729b5, you should be seeing both yaml and json unified files in the respective zip bundles.

tallytalwar committed:
Delete unified yaml or json post bundling to zip

tallytalwar commented:
Ok 19fd37e deletes the unified yaml/json file post bundling to a zip archive.

As an example on tron:

<img width="321" alt="screen shot 2018-11-01 at 10 21 09 am" src="https://user-images.githubusercontent.com/360641/47867874-0a3c8d80-ddc0-11e8-8d38-2cbd5b5da4b6.png">

tallytalwar commented:
OK tried bubble-wrap with the following import yaml:
```
import:
    - bubble-wrap-style.yaml
    - themes/bubble-wrap-icons.yaml
    - themes/label-11.yaml
```

Everything looks good with the unified bundle on ES.

bcamper commented:
Can we please use the original root scene file name for the final merged
scene file?

On Thu, Nov 1, 2018 at 1:33 PM Varun <notifications@github.com> wrote:

> Ok 19fd37e
> <https://github.com/tangrams/bundler/commit/19fd37e03dfd74fd8b6f57ef13fe7a85f5b6ea12>
> deletes the unified yaml/json file post bundling to a zip archive.
>
> As an example on tron:
>
> [image: screen shot 2018-11-01 at 10 21 09 am]
> <https://user-images.githubusercontent.com/360641/47867874-0a3c8d80-ddc0-11e8-8d38-2cbd5b5da4b6.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tangrams/bundler/pull/10#issuecomment-435117026>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABBXbMjBuSAC88celsZia-8J9RgOu2Mks5uqzBugaJpZM4XFVmD>
> .
>

tallytalwar commented:
That will over-write the original root scene file though.

bcamper commented:
> That will over-write the original root scene file though.

Can you please add an output filename (with destination path) option,? Doesn't make sense to have it hardcoded. I actually think letting it overwrite the original file is fine in some cases (not unusual for command line tools); using standard out when no file specified would also make sense.

But, for zip output (isn't that the only mode right now anyway? another thing that should be optional in future - so you can either merge or zip or both), it should be the same name as the archive anyway to ensure it loads (per our zip archive naming rules). Make sense?

tallytalwar committed:
Reuse original filename when doing unified bundling

tallytalwar commented:
Ok ee94c6b does it, original file will still be retained.
Bundled unified file will be named aptly.

tallytalwar commented:
I will add non bundling but just generating a unified merged yaml/json option to the bundler in a separate PR. This is when one will need to provide an output filename.

tallytalwar committed:
Option to only merge and generate an imported file

- this option will not create a zip archive, but just generate an output
file with everything imported
- by deault generates yaml file
- works with "-j" to generate json file

tallytalwar commented:
Ok gave in and added an option to only do merging and export a merged yaml/json file, this option will not create any zip archive. Refer: 1c48857

@bcamper Does this satisfy all requirements?
Also thanks for reviewing this and testing on JS :D.

tallytalwar committed:
Check for None draw blocks

- refill bundling was crashing because it had an empty draw block
(None).

tallytalwar committed:
Cleanup

- laye values can have a scalar (globals), only parse dict
- textures could be empty and hence will not have a url key

tallytalwar committed:
Fix uniform texture, texture reference usecase

- Was incorrectly updating the uniform texture node to have explicit
texture path set when it was using a scene texture reference, this was
causing issues with mipmap and other texture properties not being
considered
- we still need to collect all used texture resources to be included in
the assets in the zip archive

nvkelso commented:
Triple checked this and LGTM :)
